File: src/api/query.py
from fastapi import APIRouter, Depends
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from src.api.documents import get_db_session
from src.dependencies import get_embedding_client, get_retrieval_service

router = APIRouter(tags=["query"])


class QueryRequest(BaseModel):
    query: str
    top_k: int = 5


class QueryResponse(BaseModel):
    answer: str
    sources: list[dict]


@router.post("/query", response_model=QueryResponse)
async def query_documents(
    request: QueryRequest,
    session: AsyncSession = Depends(get_db_session),
):
    service = get_retrieval_service()
    embedding_client = get_embedding_client()

    async def embed(text: str) -> list[float]:
        results = await embedding_client.embed_batch([text])
        return results[0]

    result = await service.query(session, request.query, embed_fn=embed)
    return QueryResponse(answer=result.answer, sources=result.sources)


class ChunksRequest(BaseModel):
    query: str
    top_k: int = 5


@router.post("/search/chunks")
async def search_chunks(
    request: ChunksRequest,
    session: AsyncSession = Depends(get_db_session),
):
    service = get_retrieval_service()
    embedding_client = get_embedding_client()

    async def embed(text: str) -> list[float]:
        results = await embedding_client.embed_batch([text])
        return results[0]

    chunks = await service.search_chunks(session, request.query, embed_fn=embed, top_k=request.top_k)
    return {"chunks": chunks}

File: rust/src/api/query.rs
use axum::extract::State;
use axum::routing::post;
use axum::{Json, Router};
use serde::{Deserialize, Serialize};

use crate::error::AppError;

use super::state::AppState;

#[derive(Deserialize)]
pub struct QueryRequest {
    pub query: String,
    #[serde(default)]
    pub top_k: Option<usize>,
    pub memory_category: Option<String>,
}

#[derive(Serialize)]
pub struct QueryResponse {
    pub answer: String,
    pub sources: Vec<serde_json::Value>,
}

#[derive(Serialize)]
pub struct ChunkHit {
    pub chunk_id: String,
    pub doc_id: String,
    pub content: String,
    pub score: f32,
    pub section_heading: Option<String>,
    pub page_number: Option<u32>,
}

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/api/query", post(query_handler))
        .route("/api/search/chunks", post(search_chunks_handler))
}

async fn query_handler(
    State(state): State<AppState>,
    Json(body): Json<QueryRequest>,
) -> Result<Json<QueryResponse>, AppError> {
    let pool = state.pool.clone();
    let retrieval = state.retrieval.clone();
    let query = body.query.clone();
    let memory_category = body.memory_category.clone();

    let result = tokio::task::spawn_blocking(move || {
        let conn = pool.get().map_err(AppError::Pool)?;
        retrieval
            .query(&conn, &query, memory_category.as_deref())
            .map_err(|e| AppError::Llm(e.to_string()))
    })
    .await
    .map_err(|e| AppError::Llm(format!("task join error: {e}")))??;

    Ok(Json(QueryResponse {
        answer: result.answer,
        sources: result.sources,
    }))
}

#[derive(Serialize)]
struct ChunksResponse {
    chunks: Vec<ChunkHit>,
}

async fn search_chunks_handler(
    State(state): State<AppState>,
    Json(body): Json<QueryRequest>,
) -> Result<Json<ChunksResponse>, AppError> {
    let pool = state.pool.clone();
    let embed = state.embed.clone();
    let vectors = state.vectors.clone();
    let query = body.query.clone();
    let top_k = body.top_k.unwrap_or(10);

    let hits = tokio::task::spawn_blocking(move || -> Result<Vec<ChunkHit>, AppError> {
        let conn = pool.get().map_err(AppError::Pool)?;

        // Embed the query
        let query_vector = embed
            .embed_query(&query)
            .map_err(|e| AppError::Embedding(e.to_string()))?;

        // Vector search
        let vector_hits = vectors
            .search_chunks(&conn, &query_vector, top_k)
            .map_err(AppError::Db)?;

        // Load chunk metadata
        use crate::db::repositories::chunks::ChunkRepository;
        let chunk_repo = ChunkRepository::new(&conn);
        let ids: Vec<String> = vector_hits.iter().map(|h| h.id.clone()).collect();
        let chunks = chunk_repo.get_by_ids(&ids).map_err(AppError::Db)?;

        let results = chunks
            .into_iter()
            .map(|c| {
                let score = vector_hits
                    .iter()
                    .find(|h| h.id == c.id)
                    .map(|h| 1.0 - h.distance)
                    .unwrap_or(0.0);
                ChunkHit {
                    chunk_id: c.id,
                    doc_id: c.doc_id,
                    content: c.content,
                    score,
                    section_heading: c.section_heading,
                    page_number: c.page_number.map(|p| p as u32),
                }
            })
            .collect();

        Ok(results)
    })
    .await
    .map_err(|e| AppError::Llm(format!("task join error: {e}")))??;

    Ok(Json(ChunksResponse { chunks: hits }))
}
